File: packages/hydris/src/http/CookieAuthority.ts
import { createHmac, timingSafeEqual } from 'node:crypto';
import { Header } from '../constants';
import type { Request } from './Request';

export type CookieOptions = {
    maxAge?: number;
    expires?: Date;
    path?: string;
    domain?: string;
    secure?: boolean;
    httpOnly?: boolean;
    sameSite?: 'Strict' | 'Lax' | 'None';
};

export class CookieAuthority {
    private static configuredSecret: string | undefined;
    private static readonly STORE_PREFIX = 'store.';
    private static readonly STORE_MAX_BYTES = 4096;
    private static readonly UNSAFE_KEYS = new Set([
        '__proto__',
        'constructor',
        'prototype',
    ]);
    private static readonly NO_SECRET =
        'CookieAuthority: no cookie signing secret configured. Set one with CookieAuthority.secret(value) before signing or reading signed cookies.';

    static secret(value: string): void {
        CookieAuthority.configuredSecret = value;
    }

    static parse(header: string | null): Map<string, string> {
        const jar = new Map<string, string>();
        if (!header) return jar;
        for (const part of header.split(';')) {
            const eq = part.indexOf('=');
            if (eq === -1) continue;
            const name = part.slice(0, eq).trim();
            if (name === '') continue;
            jar.set(name, decodeURIComponent(part.slice(eq + 1).trim()));
        }
        return jar;
    }

    static serialize(
        name: string,
        value: string,
        options: CookieOptions = {},
    ): string {
        const parts = [`${name}=${encodeURIComponent(value)}`];
        if (options.maxAge !== undefined) {
            parts.push(`Max-Age=${Math.floor(options.maxAge)}`);
        }
        if (options.expires !== undefined) {
            parts.push(`Expires=${options.expires.toUTCString()}`);
        }
        parts.push(`Path=${options.path ?? '/'}`);
        if (options.domain !== undefined)
            parts.push(`Domain=${options.domain}`);
        if (options.secure) parts.push('Secure');
        if (options.httpOnly) parts.push('HttpOnly');
        if (options.sameSite !== undefined) {
            parts.push(`SameSite=${options.sameSite}`);
        }
        return parts.join('; ');
    }

    private static dropUnsafe(key: string, value: unknown): unknown {
        return CookieAuthority.UNSAFE_KEYS.has(key) ? undefined : value;
    }

    private static equal(a: string, b: string): boolean {
        const left = Buffer.from(a);
        const right = Buffer.from(b);
        return left.length === right.length && timingSafeEqual(left, right);
    }

    constructor(private readonly request: Request<unknown>) {}

    private requireSecret(): string {
        const secret = CookieAuthority.configuredSecret;
        if (secret === undefined || secret.length === 0) {
            throw new Error(CookieAuthority.NO_SECRET);
        }
        return secret;
    }

    private mac(name: string, value: string): string {
        return createHmac('sha256', this.requireSecret())
            .update(name)
            .update('\x00')
            .update(value)
            .digest('base64url');
    }

    sign(name: string, value: string): string {
        return `${value}.${this.mac(name, value)}`;
    }

    setCookie(
        res: Response,
        name: string,
        value: string,
        options?: CookieOptions,
    ): void {
        res.headers.append(
            Header.SetCookie,
            CookieAuthority.serialize(name, this.sign(name, value), options),
        );
    }

    get(name: string): string | null {
        const signed = this.request.cookies.get(name);
        if (signed === undefined) return null;
        const dot = signed.lastIndexOf('.');
        if (dot === -1) return null;
        const value = signed.slice(0, dot);
        return CookieAuthority.equal(
            signed.slice(dot + 1),
            this.mac(name, value),
        )
            ? value
            : null;
    }

    stores(): Record<string, unknown> {
        const seed: Record<string, unknown> = {};
        for (const [name, value] of this.request.cookies) {
            if (!name.startsWith(CookieAuthority.STORE_PREFIX)) continue;
            const key = name.slice(CookieAuthority.STORE_PREFIX.length);
            if (
                CookieAuthority.UNSAFE_KEYS.has(key) ||
                value.length > CookieAuthority.STORE_MAX_BYTES
            ) {
                continue;
            }
            try {
                const parsed = JSON.parse(value, CookieAuthority.dropUnsafe);
                if (
                    parsed !== null &&
                    typeof parsed === 'object' &&
                    !Array.isArray(parsed)
                ) {
                    seed[key] = parsed;
                }
            } catch {}
        }
        return seed;
    }
}

File: packages/hydris/Hydris/Http/CookieAuthority.cs
using System.Buffers.Text;
using System.Collections.Frozen;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Hydris.Http;

public sealed class CookieAuthority(Request request) {
    private static byte[]? SecretBytes;

    private const string NoSecret =
        "CookieAuthority: no cookie signing secret configured. Set one with CookieAuthority.Secret(value) before signing or reading signed cookies.";

    private const string StorePrefix = "store.";
    private const int StoreMaxBytes = 4096;

    private static readonly FrozenSet<string> UnsafeKeys =
        new[] { "__proto__", "constructor", "prototype" }.ToFrozenSet(StringComparer.Ordinal);

    public static void Secret(string value) =>
        SecretBytes = string.IsNullOrEmpty(value) ? null : Encoding.UTF8.GetBytes(value);

    public static IReadOnlyDictionary<string, string> Parse(string? header) {
        var jar = new Dictionary<string, string>();
        if (string.IsNullOrEmpty(header))
            return jar;

        foreach (var part in header.Split(';')) {
            var eq = part.IndexOf('=');
            if (eq < 0)
                continue;
            var name = part[..eq].Trim();
            if (name.Length == 0)
                continue;
            jar[name] = Uri.UnescapeDataString(part[(eq + 1)..].Trim());
        }

        return jar;
    }

    public static string Serialize(string name, string value, CookieOptions? options = null) {
        options ??= new CookieOptions();
        var parts = new List<string> { $"{name}={Uri.EscapeDataString(value)}" };

        if (options.MaxAge is int maxAge)
            parts.Add($"Max-Age={maxAge}");
        if (options.Expires is DateTimeOffset expires)
            parts.Add($"Expires={expires.UtcDateTime:R}");
        parts.Add($"Path={options.Path ?? "/"}");
        if (options.Domain is not null)
            parts.Add($"Domain={options.Domain}");
        if (options.Secure)
            parts.Add("Secure");
        if (options.HttpOnly)
            parts.Add("HttpOnly");
        if (options.SameSite is SameSiteMode sameSite)
            parts.Add($"SameSite={sameSite}");

        return string.Join("; ", parts);
    }

    public string Sign(string name, string value) => $"{value}.{Mac(name, value)}";

    public void SetCookie(Reply reply, string name, string value, CookieOptions? options = null) {
        ArgumentNullException.ThrowIfNull(reply);
        reply.Cookie(name, Sign(name, value), options);
    }

    public string Stores() {
        var seed = new JsonObject();
        foreach (var (name, value) in request.Cookies) {
            if (!name.StartsWith(StorePrefix, StringComparison.Ordinal))
                continue;
            var key = name[StorePrefix.Length..];
            if (UnsafeKeys.Contains(key) || value.Length > StoreMaxBytes)
                continue;

            JsonNode? parsed;
            try {
                parsed = JsonNode.Parse(value);
            } catch (JsonException) {
                continue;
            }

            if (parsed is not JsonObject entry)
                continue;
            Sanitize(entry);
            seed[key] = entry;
        }

        return seed.ToJsonString();
    }

    private static void Sanitize(JsonObject node) {
        foreach (var key in UnsafeKeys) {
            if (node.ContainsKey(key))
                node.Remove(key);
        }
        foreach (var (_, child) in node) {
            if (child is JsonObject nested)
                Sanitize(nested);
        }
    }

    public string? Get(string name) {
        if (!request.Cookies.TryGetValue(name, out var signed))
            return null;
        var dot = signed.LastIndexOf('.');
        if (dot < 0)
            return null;

        var value = signed[..dot];
        return Equal(signed[(dot + 1)..], Mac(name, value)) ? value : null;
    }

    private string Mac(string name, string value) {
        var nameBytes = Encoding.UTF8.GetBytes(name);
        var valueBytes = Encoding.UTF8.GetBytes(value);
        var buffer = new byte[nameBytes.Length + 1 + valueBytes.Length];
        nameBytes.CopyTo(buffer, 0);
        buffer[nameBytes.Length] = 0;
        valueBytes.CopyTo(buffer, nameBytes.Length + 1);

        var hash = HMACSHA256.HashData(RequireSecret(), buffer);
        return Base64Url.EncodeToString(hash);
    }

    private static byte[] RequireSecret() =>
        SecretBytes ?? throw new InvalidOperationException(NoSecret);

    private static bool Equal(string a, string b) =>
        CryptographicOperations.FixedTimeEquals(Encoding.UTF8.GetBytes(a), Encoding.UTF8.GetBytes(b));
}
